Lesson19/elifbasak/script.js

```diff
@@ -0,0 +1,165 @@
+-----------------------------------------------------------
+1. Create an `addItem` method in the `ShoppingCart` class.
+2. The method should:
+   - Accept `name`, `price`, and `quantity` as parameters.
+   - Check if the item already exists in the cart.
+     - If it exists, increase the quantity.
+     - Otherwise, add the new item to the `#items` array.
+*/
+
+/*
+-----------------------------------------------------------
+  STEP 3: Remove Items from the Cart
+-----------------------------------------------------------
+1. Add a `removeItem` method to the `ShoppingCart` class.
+2. The method should:
+   - Accept the `name` of the item to remove.
+   - Remove the item from the `#items` array if it exists.
+*/
+
+/*
+-----------------------------------------------------------
+  STEP 4: Calculate the Total Cost
+-----------------------------------------------------------
+1. Add a `getTotal` method to the `ShoppingCart` class.
+2. The method should:
+   - Calculate and return the total cost of all items in 
+     the cart.
+*/
+
+/*
+-----------------------------------------------------------
+  STEP 5: Apply a Discount
+-----------------------------------------------------------
+1. Add an `applyDiscount` method to the `ShoppingCart` class.
+2. The method should:
+   - Accept a discount code (e.g., 'SAVE10', 'SAVE20').
+   - Apply a percentage discount to the total cost if the 
+     code is valid.
+3. Use an object to store discount codes and their values.
+*/
+
+class ShoppingCart {
+  #items;
+
+  constructor() {
+    this.#items = [];
+  }
+
+  viewCart() {
+    console.log("Viewing cart....");
+    for (const item of this.#items) {
+      console.log(item);
+    }
+  }
+
+  addItem(name, price, quantity) {
+    console.log("Adding new item to the cart...");
+    for (const item of this.#items) {
+      if (name === item.name) {
+        console.log("Item found in the cart, increasing the quantity.");
+        item.quantity++;
+        return;
+      }
+    }
+
+    const id = new Date().getMilliseconds();
+    this.#items.push({ id, name, price, quantity });
+  }
+
+  // In real life, we usually use id to find elements or do any manipulations
+  removeItem(name) {
+    console.log("Removing the item...");
+    for (let i = 0; i < this.#items.length; i++) {
+      const currentElement = this.#items[i];
+      if (name.toLowerCase() === currentElement.name.toLowerCase()) {
+        this.#items.splice(i, 1);
+        return;
+      }
+    }
+  }
+
+  getTotal() {
+    console.log("Calculating the total cost of the cart....");
+    let total = 0;
+    for (const item of this.#items) {
+      total += item.quantity * item.price.amount;
+    }
+    console.log(`Total amount of the cart is ${total} TRY.`);
+    return total;
+  }
+
+  applyDiscount({ discountCode, discountPercentage }) {
+    console.log("Applying discount....");
+    const discountCodes = { SAVE10: 0.1, SAVE20: 0.2 };
+    if (typeof discountCode !== "string") {
```

Comment:
Error validation here can me much simplier: 

`if (!discountCodes[discountCode]) {`. If the discountCode matches keys available in the discountCodes object, the discount can be used, otherwise return total value.